### ISSUE TITLE
Range types for integers (or refinement types?)

### ISSUE THREAD

**Issue by BryanQuigley** [rust-lang/rust#19801]
It seems like a natural extension of how variables (immutable by default, mutable if specified) are defined to allow the programmer to dictate a specific range of allowed values for an integer.  If I know a value is only valid between 0-1000 the sooner I declare that the better it is for catching bugs, off by one errors, and more...  

I'm not sure what exact syntax would work, maybe:
```rust
let mut(0,1000) x = 0i;
```
x is only valid from 0-1000 inclusive.

(Apologies if this is already possible, I've been parsing the docs trying to learn Rust.)

**Comment by kmcallister** [comment]
Ada has this. It would be great in Rust, as we're going for the same niche of low-level + high-assurance.

I wonder though if we should go for a more general system, such as refinement types or pre/post-conditions.  There have been a number of successes bolting these onto languages like [Haskell](http://goto.ucsd.edu/~rjhala/liquid/haskell/blog/about/), [F#](http://research.microsoft.com/en-us/projects/f7/), and [C#](http://rise4fun.com/specsharp/).  aiui, they manage this without heavy changes to the core language.  The condition checker is a mostly-separate tool that gathers proof obligations from the source code and passes them to a SMT solver.

Basically I think this is an area where we should give researchers and other interested parties some time to experiment before we standardize something into Rust itself.

**Comment by darinmorrison** [comment]
Something like this is in fact possible now with the associated types machinery. I have a [library](http://epsilonz.github.io/shoggoth.rs/doc/shoggoth/ty/index.html) that implements type-level binary natural numbers (and soon ints and rationals). There are some ergonomic issues with using it still, but some of that can be alleviated with macros.

For example, I have a local rust branch that supports macro expansion in types (need to submit a PR), so you can write things like this and it works:

```rust
fn pop<A, N: Tm<Nat>>(xs: NatVec<A, nat!(N + 1n)>) -> (A, NatVec<A, N>) {
    …
}
let xs: NatVec<A, nat!(512n)> = NatVec::<nat!(512n)>::new();
let ys: NatVec<A, nat!(511n)> = pop(xs).1;
```

**Comment by steveklabnik** [comment]
I'm pulling a massive triage effort to get us ready for 1.0. As part of this, I'm moving stuff that's wishlist-like to the RFCs repo, as that's where major new things should get discussed/prioritized.

This issue has been moved to the RFCs repo: rust-lang/rfcs#671
